runelite-client/src/main/java/net/runelite/client/plugins/twitch/TwitchPlugin.java

```diff
@@ -0,0 +1,420 @@
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package net.runelite.client.plugins.twitch;
+
+import com.google.common.eventbus.Subscribe;
+import com.google.inject.Provides;
+import javax.inject.Inject;
+import net.engio.mbassy.listener.Handler;
+import net.runelite.api.ChatMessageType;
+import net.runelite.api.GameState;
+import net.runelite.api.Varbits;
+import net.runelite.api.events.CommandExecuted;
+import net.runelite.api.events.ConfigChanged;
+import static net.runelite.client.callback.Hooks.log;
+import net.runelite.client.chat.ChatMessageManager;
+import net.runelite.client.config.ConfigManager;
+import net.runelite.client.plugins.Plugin;
+import net.runelite.client.plugins.PluginDescriptor;
+import org.apache.commons.lang3.StringUtils;
+import org.kitteh.irc.client.library.Client;
+import org.kitteh.irc.client.library.event.channel.ChannelMessageEvent;
+import org.kitteh.irc.client.library.feature.twitch.TwitchDelaySender;
+import org.kitteh.irc.client.library.feature.twitch.TwitchListener;
+import org.kitteh.irc.client.library.feature.twitch.event.RoomStateEvent;
+import org.kitteh.irc.client.library.feature.twitch.event.UserNoticeEvent;
+import org.kitteh.irc.client.library.feature.twitch.event.UserStateEvent;
+
+@PluginDescriptor(
+		name = "Twitch",
+		enabledByDefault = false
+)
+public class TwitchPlugin extends Plugin
+{
+	@Inject
+	private net.runelite.api.Client client;
+
+	@Inject
+	private TwitchConfig config;
+
+	@Inject
+	private ChatMessageManager chatMessageManager;
+
+	private Client irc;
+
+	private boolean subscribed;
+
+	private boolean mod;
+
+	private boolean broadcaster;
+
+	private boolean subsOnly;
+
+	private boolean emoteOnly;
+
+	private String badges = "";
+
+	@Provides
+	TwitchConfig provideConfig(ConfigManager configManager)
+	{
+		return configManager.getConfig(TwitchConfig.class);
+	}
+
+	@Override
+	protected void startUp()
+	{
+		twitchShutdown(irc);
+		irc = twitchSetup(config.oauthToken(), config.channelName(), config.nickName());
+	}
+
+	@Override
+	protected void shutDown()
+	{
+		twitchShutdown(irc);
+	}
+
+	@Subscribe
+	public void onConfigChanged(ConfigChanged event)
+	{
+		if (event.getGroup().equals("twitch") && (event.getKey().equals("channel") || event.getKey().equals("oauth") || event.getKey().equals("nick")))
+		{
+			twitchShutdown(irc);
+			irc = twitchSetup(config.oauthToken(), config.channelName(), config.nickName());
+		}
+	}
+
+	public boolean isChatboxTransparent()
+	{
+		return client.isResized() && client.getVar(Varbits.TRANSPARENT_CHATBOX) == 1;
+	}
+
+	public Client twitchSetup(String o, String nick, String name)
```

Comment:
o -> oauth

Reply:
Also what happens when bad input is sent (like an empty space, like the default config, or a non existent token/user)? How is that handled? I'm guessing the TwitchListener class handles it, but how is the user notified of errors? (I haven't tested this plugin yet)

Reply:
TwitchListener does handle that stuff. If there is an invalid oauth token the message won't be sent at all and won't show up in the in game chat. But if an incorrect nickname is entered with a correct oauth token then the message will be sent and show up in twitch chat. The only issue this causes is that the ingame chat will show the wrong username.

Eg.
![image](https://i.gyazo.com/1eae81cb49bd4d4948866844b74d9687.gif)

Reply:
Currently there's no error given for this